sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala

```diff
@@ -56,39 +64,94 @@ object CommandUtils extends Logging {
-        val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
-        val pathFilter = new PathFilter with Serializable {
-          override def accept(path: Path): Boolean = isDataPath(path, stagingDir)
+      val sizeWithDeserFactorsForPartitions =
+        if (sessionState.conf.parallelFileListingInStatsComputation) {
+          val paths = partitions.map(x => new Path(x.storage.locationUri.get))
+          val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
+          val pathFilter = new PathFilter with Serializable {
+            override def accept(path: Path): Boolean = isDataPath(path, stagingDir)
+          }
+          val deserFactCalcEnabled = sessionState.conf.deserFactorStatCalcEnabled
+          val hadoopConf = sessionState.newHadoopConf()
+          val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(
+            paths, hadoopConf, pathFilter, spark, areRootPaths = true)
+          fileStatusSeq.flatMap { case (_, fileStatuses) =>
+            fileStatuses.map { fileStatus =>
+              sizeInBytesWithDeserFactor(
+                deserFactCalcEnabled,
+                hadoopConf,
+                fileStatus,
+                catalogTable.storage.serde)
+            }
+          }
+        } else {
+          partitions.map { p =>
+            calculateLocationSize(
+              sessionState,
+              catalogTable.identifier,
+              p.storage.locationUri,
+              p.storage.serde)
+          }
         }
-        val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(
-          paths, sessionState.newHadoopConf(), pathFilter, spark, areRootPaths = true)
-        fileStatusSeq.flatMap(_._2.map(_.getLen)).sum
-      } else {
-        partitions.map { p =>
-          calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-        }.sum
-      }
+      sumSizeWithMaxDeserializationFactor(sizeWithDeserFactorsForPartitions)
     }
     logInfo(s"It took ${(System.nanoTime() - startTime) / (1000 * 1000)} ms to calculate" +
       s" the total size for table ${catalogTable.identifier}.")
     totalSize
   }
 
+  def sumSizeWithMaxDeserializationFactor(
+      sizesWithFactors: Seq[SizeInBytesWithDeserFactor]): SizeInBytesWithDeserFactor = {
+    val definedFactors = sizesWithFactors.filter(_.deserFactor.isDefined).map(_.deserFactor.get)
+    SizeInBytesWithDeserFactor(
+      sizesWithFactors.map(_.sizeInBytes).sum,
+      if (definedFactors.isEmpty) None else Some(definedFactors.max))
+  }
+
+  def sizeInBytesWithDeserFactor(
+      calcDeserFact: Boolean,
+      hadoopConf: Configuration,
+      fStatus: FileStatus,
+      serde: Option[String]): SizeInBytesWithDeserFactor = {
+    assert(fStatus.isFile)
+    val factor = if (calcDeserFact) {
+      val isOrc = serde.contains(orcSerDeCanonicalClass) || fStatus.getPath.getName.endsWith(".orc")
```

Comment:
Are there any plan to support datasource table to calculate this stats on demand?  Maybe we can also add an option to enable this for on demand calculation?

Reply:
I have not thought about it yet but I would like to keep the scope of this PR minimal. When this is in we can come back to it.